Adding SIG Process/Workflow Guide and related images (#5565)
kubernetes/community: contributors/chairs-and-techleads/process-workflow-guide.md

For brevity’s sake, not every aspect of running a SIG, WG, or subproject is covered. If you have questions, please reach out in [#chairs-and-techleads](https://kubernetes.slack.com/messages/chairs-and-techleads).

## Ways of Working Agreement

This one comes first as it’s a helpful way to summarize how your SIG, WG, or subproject operates in terms of communication, processes, and roles and responsibilities. This can serve as a useful onboarding tool or help you prevent process drift or abandonment. 

### Example: SIG Release/Release Engineering

The [Release Engineering](https://github.com/kubernetes/sig-release/tree/master/release-engineering) subproject of SIG Release has published a Ways of Working agreement that covers communication channels and how they’re used; roles and responsibilities; how they discuss work; and other major aspects of workflow and process. 

### Making the Most out of Meetings

Most, if not all, SIGs, WGs, and subprojects already have templates for organizing bimonthly meetings. [Here is a collection of templates](https://miro.com/templates/team-meetings/) for remote meetings that you might also find useful.

> You’ll find some SIG-specific meeting tips in Release Engineering’s WoW agreement. A few more:

Reviewer: What is "Release Engineering’s WoW agreement"? Is there a link?

Author: There's a link earlier in the guide -- WoW = "Ways of Working"

Reviewer: oh, that makes sense. I didn't connect the abbreviation to the previous text. sorry

Author: No problem :)

- [Ways of Working Agreement](#ways-of-working-agreement) (SIG Release/Release Engineering)
  - [Example: SIG Release/Release Engineering](#example-sig-releaserelease-engineering)
- [Making the Most out of Meetings](#making-the-most-out-of-meetings) (SIGs Instrumentation, Release, Storage, Auth)
- [Triage and Triaging Tools](#triage-and-triaging-tools)
- [Triage Meetings](#triage-meetings) (SIGs API Machinery, Instrumentation)
  - [Staffing Triage Activities](#staffing-triage-activities) (SIG Docs)
- [Triage Party](#triage-party) (SIGs Cluster Lifecycle, CLI, Release)
  - [Example: Triage Party for Skaffold](#example-triage-party-for-skaffold) (SIG Cluster Lifecycle)
- [GitHub Project Boards](#github-project-boards) (SIGs Node, Windows, Release) 
  - [Board Set-up Basics](#board-set-up-basics)
  - [Example: Pull Request Triage Board](#example-pull-request-triage-board-sig-node) (SIG Node)
  - [Example: Board Highlighting Risks](#example-board-highlighting-risks-sig-windows) (SIG Windows) 
  - [Example: KEP-Specific Issue Triage](#example-kep-specific-issue-triage-sig-release) (SIG Release)

> - Triaging CI Signal Needs (SIG Windows)

Reviewer 2: I'm not sure this item exists later on :x

# Process and Workflow Guide

Author: **[@LappleApple](https://github.com/lappleapple)** || Contributors: **[@annajung](https://github.com/annajung)**, **[@kikisdeliveryservice](https://github.com/kikisdeliveryservice)**, **[@neolit123](https://github.com/neolit123)** 

## Who This Is For

> - SIG leads who want guidance on how to structure workflow and processes using models explored successfully by other SIGs 

Reviewer 2: ```suggestion
- Leads who want guidance on how to structure workflow and processes using models explored successfully by other groups 
```
Might be nitty, but SIGs are our main group, but this doc would be applicable for WGs and subproject. Language would need to be updated throughout the doc though.

- Contributors who want to support SIGs with process and workflow improvements
- Contributors who want to become SIG leads or drive Kubernetes Enhancements Proposals (KEPs) and projects

## Why This Exists

The primary reason is to reduce risk. Without processes in place to engage and graduate contributors, keep work flowing forward to completion, delegate, and ensure that the highest-value work gets done, Kubernetes will lose contributors, momentum, and growth opportunities. And you, as leads, might burn out from trying to do it all. Process assists in delegation, which is how overworked and overstretched SIG leads can load-balance. 

The idea here is progress, not perfection. Please try to put past process experiments and things-that-didn’t-work behind you. What failed yesterday might work today. You might need to recalibrate your idea of “failure”—a small bit of progress is still progress. Also, from failures we learn.

> You might also take a cue from SIG Release and create a Program Manager role for your SIG. Reach out to @lauri if you’d like to explore this option.

Reviewer 3: ```suggestion
You might also take a cue from SIG Release and create a Program Manager role for your SIG. Reach out to [@LappleApple](https://github.com/lappleapple) if you’d like to explore this option.
```

  - [Tips for Handling KEPs That Require Cross-SIG Coordination](#tips-for-handling-keps-that-require-cross-sig-coordination)
  - [Managing KEP Scope](#managing-kep-scope)
  - [When You Need More Help with KEPs/KEP Process](#when-you-need-more-help-with-kepskep-process)
- [Creating Roadmaps](#creating-roadmaps)
  - [Example: Roadmapping Templates for a Release Cycle](#example-roadmapping-templates-for-a-release-cycle-sigs-docs-release) (SIGs Docs, Release)
  - [Example: Brainstorming Template for Creating a Roadmap](#example-brainstorming-template-for-creating-a-roadmap-sig-releaserelease-engineering) (SIG Release/Release Engineering)
  - [Example: Umbrella Issues for Roadmapping an Initiative/Epic](#example-umbrella-issues-for-roadmapping-an-initiativeepic-sigs-testing-release) (SIGs Testing, Release)
- [Managing Communications: Delegating and Driving Transparency](#managing-communications-delegating-and-driving-transparency)
  - [Slack](#slack)
  - [Mailing List and GitHub Notification Best Practices](#mailing-list-and-github-notification-best-practices) (SIG Contributor Experience)

> For brevity’s sake, not every aspect of running a SIG is covered. If you have questions, please reach out in **#chairs-and-leads**.

Reviewer 2: ```suggestion
For brevity’s sake, not every aspect of running a SIG is covered. If you have questions, please reach out in [#chairs-and-leads](https://kubernetes.slack.com/messages/chairs-and-techleads).
```

Added link 👍

Meetings provide an opportunity to delegate work to contributors and to make group decisions about KEP backlogs, future plans, roadmaps, etc. For this reason, consider keeping discussions about specific topics brief and high-level. The async sidebar option (discussion on a GitHub issue, or a Slack thread) or small-group call is often better for working out details on a specific topic in order to make the most out of group meeting time.

## Triage and Triaging Tools

Triaging is an effective method for improving SIG responsiveness to contributor requests—and, consequently, keeping contributors engaged and motivated. It’s also useful for:
- keeping on top of incoming work and delegating it to contributors
- maintaining company agnosticism. “Devs from other SIGs don’t have to ask for any favors in getting PRs to move forward—it’s just part of the process,” API Machinery Chair Fede Bongiovanni says. “Features don't get biased development.”

> You can find [comprehensive triaging guidelines](https://github.com/kubernetes/community/blob/c78c26294b8bd137dd5cc0c807aadc728dad6a57/contributors/guide/issue-triage.md) here, along with information about triage tools used across the project:

Reviewer 2: ```suggestion
You can find [comprehensive triaging guidelines](/contributors/guide/issue-triage.md) here, along with information about triage tools used across the project:
```

For links to items in the same repo, we prefer to use the path (also not linking to a specific commit)

- **Gubernator**: a personal dashboard that shows you PRs awaiting your review
- **Triage Party**: a stateless web app to optimize issue and PR triage for large open-source projects using the GitHub API. More details are below. (Cluster Lifecycle, SIG Release) 

> - **GitHub project boards**: these kanban boards allow you to track GitHub issues and PRs in an automated way. More details are below.

Author: Good suggestion -- just added links

### Triage Meetings

Triaging is highly recommended for improving workflow and responsiveness to contributors. 

> Some SIGs, like API Machinery and [Instrumentation](https://docs.google.com/document/d/1FE4AQ8B49fYbKhfg4Tx0cui1V0eI4o3PxoqQPUwNEiU/edit), host triage-specific meetings to manage their backlogs. If your SIG’s backlog is small, however, then you can triage and refine during your bimonthly meetings by setting aside time. In November 2020 we rolled out a needs-triage label to support triage activities. You can add your [user feedback about the label here](https://github.com/kubernetes/community/issues/5299#issuecomment-776914413).

Reviewer 2: ```suggestion
Some SIGs, like API Machinery and [Instrumentation](https://docs.google.com/document/d/1FE4AQ8B49fYbKhfg4Tx0cui1V0eI4o3PxoqQPUwNEiU/edit), host triage-specific meetings to manage their backlogs. If your SIG’s backlog is small, however, then you can triage and refine during your bimonthly meetings by setting aside time. In November 2020 we rolled out a needs-triage label to support triage activities. You can add your [user feedback about the label here](https://github.com/kubernetes/community/issues/5299).
```

SIG Docs rotates the triage facilitator role with its “[wrangler of the week](https://github.com/kubernetes/website/wiki/PR-Wranglers)” schedule. The wrangler balances their time between reviewing PRs and triaging new issues. 

### Triage Party 

#### Example: Triage Party for Skaffold

[Triage Party](https://github.com/google/triage-party) is an open-source tool that focuses on “reducing response latency for incoming GitHub issues and PR's, and ensure[s] that conversations are not lost in the ether.” Project creator Thomas Strömberg reports that using Triage Party for Minikube and Skaffold workflows has led to dramatic improvements in responsiveness to contributor requests. By “dramatic,” he means *40 times faster*.

<img src="triage_party_skaffold.png" width="60%" height="60%">

**Other Uses of Triage Party**:
- SIG CLI uses it to manage [kubectl](https://tp.kuberneddies.dev/s/daily) workflow, including quarterly bug scrubs and daily/weekly triage

> - SIG Release has set up a Triage Party instance and will begin using it for bug triage, issue triage, and other activities in 1.21.

Reviewer 2: Is this live now?

Author: Actually yes, it went live a couple of weeks ago: https://release.triage.k8s.io/s/milestone. I'll add that link.

- Alert the other SIGs well ahead of time of your needs. You might need to plan a cycle in advance for an especially complex KEP.
- Develop a coordination plan that is self-service and lightweight—a simple spreadsheet or GitHub tracking issue, highlighting what is needed from each SIG and by when, can help you stay on top of progress and upcoming needs.
- If you’ve not had time to discuss the KEP with participating SIGs, try reducing the scope of the KEP for the upcoming release so that it doesn't affect other SIGs. This way, fewer people from other SIGs need to give tech reviews for KEP or code.

#### Managing KEP Scope

Breaking down large KEPs into smaller increments of value helps you to ship faster and more frequently, and creates shorter feedback loops. 

Network and Storage are two SIGs that regularly break down large KEPs to ship more predictably. An example from Storage is volume health: The alpha version was a very small feature that could be expanded in later cycles. When it’s not possible or simple to break down KEPs, consider holding them until the next release cycle to avoid missing the Code Freeze deadline.

#### When You Need More Help with KEPs/KEP Process

> Please ping the Enhancements subproject at **#enhancements** with your questions, needs, and concerns. In addition to having extensive knowledge about the KEP process and how to manage it, the subproject also includes several product manager-contributors who will embed with your SIG and help you with prioritization, user story creation for KEPs, roadmapping, and other efforts that can save you time and reduce complexity.

Reviewer 2: ```suggestion
Please ping the Enhancements subproject at [#enhancements](https://kubernetes.slack.com/messages/enhancements) with your questions, needs, and concerns. In addition to having extensive knowledge about the KEP process and how to manage it, the subproject also includes several product manager-contributors who will embed with your SIG and help you with prioritization, user story creation for KEPs, roadmapping, and other efforts that can save you time and reduce complexity.
```

- **Who** is the customer (user personas, specific parties served)
- **What** is the deliverable (including milestones, or increments of value that roll up to the deliverable)
- **Why** is it needed/urgent—what is its purpose? (including metrics and objectives) 
- **When** is it needed (what opportunities will we lose, by when, if we don’t do this now?)
- **Where** will the work be done: SIGs, components, areas of architecture 
- **Acceptance Criteria**: how do we know when we’re “done”?
- **Risks**: of not doing it, and the risks we might face during development 
- **What’s out of scope**

### Example: Brainstorming Template for Creating a Roadmap (SIG Release/Release Engineering)

Release Manager Adolfo “Puerco” García Veytia and SIG Release Tech Lead Dan Mangum used Miro to brainstorm a roadmap for creating a bill of materials:

<img src="BOM board.png">

SIG Release/Release Engineering is using [this template](https://docs.google.com/document/d/1U-SxIQKwEDmlj3QMxmKUl2laFRKaelDwBBBqWtnxPvw/edit) to draft a 2021 North Star vision that tells a coherent story to the project and contributors about what work is important this year, and why.

> SIG Docs also does quarterly planning and objectives... (**find out more**)

Reviewer 2: ```suggestion
SIG Docs also does quarterly planning and objectives... (**find out more**)
```
is this a TODO for the future?

Author: Oops --removed

Here’s [another good example](https://github.com/kubernetes/test-infra/issues/18549) of an umbrella issue, from SIG Testing:

- Describes at the top why the work is important
- Includes a section for decisions to be made during development
- Includes a checklist to track supporting tasks 

## Managing Communications: Delegating and Driving Transparency

Across a project the size of Kubernetes, communications can easily be lost or missed. Many contributors receive literally hundreds of GitHub notifications a day. Many notes get pushed to Slack, and SIG Chairs and Leads face many direct messages on top of channel pings. Time to read important but long documents can be scarce. 

### Slack

- Use direct messages only for sensitive topics (personal, HR-related).  
- When someone pings via DM about work or a question that could be discussed/answered by others, direct them to your public channel.
- Use 2-3 channels as your main go-to for discussion.
- Create team handles and use them to draw attention to time-sensitive or workflow-related Slack messages (posted GitHub issues, PRs). This can drive delegation and discussion.

### Mailing List and GitHub Notification Best Practices

> Visit /community for [documentation on this topic](https://github.com/kubernetes/community/blob/master/communication/best-practices.md).

Reviewer 2: ```suggestion
See the [communication best practices for mailing list and GitHub notifications](/communication/best-practices.md).
```

The doc is in k/community already 👍
